### ISSUE TITLE
Capture on "Completed" instead of on "Processing"

### ISSUE THREAD

**Issue by eivindlm** [vippsas/vipps-woocommerce#62]
Currently it appears like the plugin captures the amount when I change order status from "Pending" to "Processing". However, as far as I have understood, I am not allowed by Norwegian law to capture the amount until I actually ship the order. So the current plugin makes me completely unable to use the "Processing" state of the order.

A more concrete example: I currently have 250 orders "Pending", of which I want to start processing 100 of them for shipping in a few days. WooCommerce lets me set the state of the orders to "Processing" (very useful to know which orders are started and which are not), but this would imply that the amount is captured, and I should not capture the amount until I ship the order.

So this issue is a request for a new setting:

    Capture amount when order is set to: {Processing,Completed}

Or actually, I cannot really see the case when the amount would need to be captured on state "Processing" instead of on "Completed", so I think you should consider simply moving the capturing permanently from "Processing" to "Completed".

Or am I missing something here?

**Comment by iverok** [comment]
This has been covered quite a bit on wp.org and elsewhere. **TLDR: You can set your preference here in the settings by choosing "Order status on return from Vipps" to "Processing". Capture will then be done on "Complete" instead.**

There is one caveat: it is *theoretically* possible that capture fails, even though reserve succeeded. In that case, you will want to halt the sending of that order until payment is ok.  However, this has to our knowledge happened just  a single time in several years of service on a lot of stores; so it is likely that this will basically never happen.

Background for this: When implementing this originally, the documentation available by WooCommerce for the reserve/capture flow for payment gateways used the "on-hold" status for orders with reserved and not captured payment.  Basically, the "Processing" status is then used for all processing of the order including shipping, while "Complete" is just the finished order status (after shipping etc). In addition, the possibility of capture failing on "complete" would have been a real problem; and we did in fact get a report of a single incident of this sort.

Unfortunately, WooCommerce also uses the "on hold" status to indicate "there is a problem with the payment of this order", and it is common to send out emails to that effect when an order is set to that status.  Additionally, we'd really like to be able to use "processing" also for stuff like partial capture and so forth, so using "on-hold" is pretty inconvenient.  Adding an extra order status between processing and capture would not be good for stores having their own order status workflow (e.g. with "shipping"). And there is no separate "payment status" unfortunately.

Because of the available documentation and the very slight risk of problems with "Complete", the current default was reluctantly implemented.

To be 100% sure to avoid the safety issue, you could either add a new order status (like "Shipping"), set that as a "capturing" status using filters, and use "processing" to as the return status; or do the capturing manually or automatically before setting the status to complete. 

As a final note: If your order is all "virtual" and "downloadable" products, the payment will be captured at once; and normally the order status will go directly to "complete".

**Comment by eivindlm** [comment]
Thank you for the quick response!

I am aware of the existing option "Order status on return from Vipps". However this will make the "Processing" stage in woocommerce equally useless for me. I want incoming orders to be "Pending", in order to utilize the "Processing" stage for the orders which have actually started processing.

I think I understand the caveat about reserve->capture, so I agree now that you should not have "Completed" as the default and only stage for capture as I suggested before.

But wouldn't it make sense to make it available as an option? Users selecting "Finished" as the capture stage in a new option would not run into any larger risk that users who select "Processing" as the return stage from vipps in the currently available option (where capture will be done on finished anyway). So having two different option boxes could make the system fit more workflows without adding any new risks.

So below are the new option that i suggest, in addition to the old. The same warning (perhaps slightly updated) about the risk involved with reserve->capture should be printed along with them:

    Order status on return from Vipps: {Pending, Processing}
    Order status for capture: {Processing, Finished}

For my workflow, I would select Pending on the first option and Finished on the second option. Others may still select "Processing" in the first option and "Finished" in the second option, in order to replicate the current behaviour. A third option may be possible, to select "Processing" on both options. This would imply that the amount will be captured immediately, but still leaving the "Finished" stage available for the user workflow. I think this also sounds like a useful possibilty.

**Comment by iverok** [comment]
"Pending" is problematic : It's used for orders begun but not completed, and I'm pretty sure WooCommerce will delete dem automatically after some time if the status isn't changed to something else. Because of this, it can't be used as a "payment complete" status - sorry about that; it's used as an indicator of "payment in process" in the Vipps plugin too. Maybe you mean "On-hold" here?

If not, it would be better to use a custom order status before "processing" instead of "Pending".

I may be getting this wrong, but would an OK flow for you be order status "on-hold" (or custom) after return from Vipps, but with no payment capture on "Processing"? Then payment capture either manually or on "Complete"?

If so, you can for now implement this using this filter:

   ` $captured_statuses = apply_filters('woo_vipps_captured_statuses', array('processing', 'completed'));`

Just remove "processing" from the array and return the statuses you want to capture on.

If you want to use custom order statuses, you will also need this:

    `$newstatus = apply_filters('woo_vipps_after_vipps_order_status', $defaultstatus, $order);`

- to have this as the return status from Vipps. 

We try to keep the number of options as low as possible (there are already a lot!) but we could perhaps add a "don't capture on Processing" checkbox to the 'advanced' section. Most of our requests on this goes in the other direction, with trying to avoid "on-hold" and having capture done earlier rather than later.

**Comment by cloveras** [comment]
Hallo fra Vipps

Jeg tenkte bare å tipse om at vi har skrevet litt om "reserve capture" og "direct capture" i FAQ-en for API-et vårt, i tilfelle det er interessant: https://github.com/vippsas/vipps-ecom-api/blob/master/vipps-ecom-api-faq.md#what-is-the-difference-between-reserve-capture-and-direct-capture

Christian, Vipps

**Comment by eivindlm** [comment]
Yes, sorry, i meant "On-Hold" and not "Pending" everywhere in my above comment. Sorry about that!

An option for "don't capture on processing" sounds excellent, and will totally cover my workflow.

**Comment by iverok** [comment]
In the mean time you should be able to write this to a file in your "plugins" directory and activate it to avoid having "process" capture funds:

```
?php
/**
 * Plugin Name: Vipps: Do not capture on process
 * Description: Ensure funds paid with Vipps only be captured on the "complete" status.
 * Version: 1.0
 */

if ( ! defined( 'ABSPATH' ) ) {
        exit;
}

add_action('plugins_loaded', function () {
  add_filter('woo_vipps_captured_statuses', function ($statuses)  {
    return array('completed');
  });
}, 5);
```
